linkd-lib/src/args.rs

```diff
@@ -0,0 +1,298 @@
+                    .parse()
+                    .map_err(|e: crypto::peer::conversion::Error| e.to_string())?;
+
+                Ok(Self {
+                    addr: addr.to_string(),
+                    peer_id,
+                })
+            },
+            None => Err("missing peer id".to_string()),
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, StructOpt)]
+pub enum Signer {
+    /// Construct signer from a secret key.
+    Key,
+    /// Connect to ssh-agent for delegated signing.
+    SshAgent,
+}
+
+impl Default for Signer {
+    fn default() -> Self {
+        Self::SshAgent
+    }
+}
+
+impl fmt::Display for Signer {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let ty = match self {
+            Self::Key => "key",
+            Self::SshAgent => "ssh-agent",
+        };
+
+        write!(f, "{}", ty)
+    }
+}
+
+impl FromStr for Signer {
+    type Err = String;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        match input {
+            "key" => Ok(Self::Key),
+            "ssh-agent" => Ok(Self::SshAgent),
+            _ => Err(format!("unsupported signer `{}`", input)),
+        }
+    }
+}
+
+#[derive(Debug, Default, Eq, PartialEq, StructOpt)]
+pub struct KeyArgs {
+    /// Location of the key file on disk.
+    #[structopt(
+        long = "key-file-path",
+        name = "key-file-path",
+        default_value = "",
+        parse(from_str),
+        required_if("signer", "key"),
+        required_if("key-source", "file")
+    )]
+    pub file_path: PathBuf,
+    /// Format of the key input data.
+    #[structopt(
+        long = "key-format",
+        name = "key-format",
+        default_value,
+        required_if("signer", "key")
+    )]
+    pub format: KeyFormat,
+    /// Specifies from which source the secret should be read.
+    #[structopt(
+        long = "key-source",
+        name = "key-source",
+        default_value,
+        required_if("signer", "key")
+    )]
+    pub source: KeySource,
+}
+
+#[derive(Debug, Eq, PartialEq, StructOpt)]
+pub enum KeyFormat {
+    Base64,
+    Binary,
+}
+
+impl Default for KeyFormat {
+    fn default() -> Self {
+        Self::Binary
+    }
+}
+
+impl fmt::Display for KeyFormat {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let source = match self {
+            Self::Base64 => "base64",
+            Self::Binary => "binary",
+        };
+        write!(f, "{}", source)
+    }
+}
+
+impl FromStr for KeyFormat {
+    type Err = String;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        match input {
+            "base64" => Ok(Self::Base64),
+            "binary" => Ok(Self::Binary),
+            _ => Err(format!("unsupported key format `{}`", input)),
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, StructOpt)]
+pub enum KeySource {
+    File,
+    InMemory,
+    Stdin,
+}
+
+impl Default for KeySource {
+    fn default() -> Self {
+        Self::Stdin
```

Comment:
I'd think `File` is a more sensible default, otherwise we block on stdin and it's confusing

Reply:
In fact not sure why you'd want stdin if you have file

Reply:
Sometime in the past, someone in an outburst of productivity produced overnight a critical piece of infrastructure. To aid the network in bootstrapping and bring replication to the plebs, born was the SEED. As the creator believed in composability and drank from the firehose called the UNIX pipe, they made it so that you can pipe the most private of private things - the secret key -  into the SEEDs stdin. It was glorious, but many an operator lost their sleep over the breaking systemd services, a lot of bash was written in those days (dark times indeed). And because of that legacy we have support for STDIN.

In fact, I thought I never say that, it has proven handy for configuration of the whale containers. There you often share the same image for different peers, so you can't really store a single key on disk as they would share it. Changing the default to file is not an outrageous thought.

Reply:
Epic